_posts/2022-03-15-turn-based-offline-rl.md

```diff
@@ -0,0 +1,344 @@
+This approach would be the way to go in the majority of the cases and is the closest to the real case example.
+Suppose you don’t know anything about the environment you are going to interact with: the best you can do is to
+perform random exploration!
+
+![Figure 4](/img/articles/2022-03-15-turn-based-offline-rl/random_policy_heatmap_reduced.png)
+*Fig.4 — State-action visitation heatmap generated by the random policy. The agent performs random exploration. As a
+result of the random behaviour, cells in the surroundings of the initial state are visited on average more often than
+further cells.*
+
+Since we start from the “S” cell at every episode, we have the highest probability of visiting the “S” state and all
+its surroundings. As we go further from it, the agent will start to pick different states depending on the run,
+and thus values on farther cells start to normalize and approach 0.0.
+
+In the following, we will describe the algorithm in details, and we will make use of these visualizations to understand
+if the turn-based approach is beneficial for learning a good policy when starting from a random one.
+
+### Algorithm
+
+Now let’s dive into the algorithm itself. Recalling the steps indicated in the previous section,
+we can describe the turn-based learning algorithm with the following pythonic pseudocode:
+
+```python
+def run_turn_based_algorithm(init_policy,
+                             num_turns,
+                             num_seeds,
+                             dataset_size,
+                             num_iters):
+    offline_dataset = []
+    current_policy = init_policy
+    num_of_trajectories_per_turn = dataset_size / num_turns
+
+    for turn in range(num_turns):
+        runs = []
+        for seed in range(num_seeds):
+            temp_dataset = offline_dataset.copy()
+            trajectories_from_new_policy = deploy_and_sample(current_policy, num_of_trajectories_per_turn)
+            temp_dataset.extend(trajectories_from_new_policy)
+            policy, performances = run_offline_rl_algorithm(temp_dataset, num_iters)
+            runs.append((policy, performances, trajectories_from_new_policy))
+
+        best_policy, best_trajectories = find_best_run(runs)
+        current_policy = best_policy
+        offline_dataset.extend(best_trajectories)
+```
+
+Let’s explain each step involved in the algorithm.  First, let’s define what are the main parameters expected by
+the algorithm:
+
+* `init_policy` — it’s the starting policy, most likely the random policy.
+* `num_turns` — this is simply the total number of turns for which you will run the algorithm.
+* `num_seeds` — if you work in RL you will be familiar with this argument: RL algorithms
+(and especially Offline RL ones) present large variability in the results due to their stochastic nature.
+That’s why instead of having one single run of the Offline RL algorithm,
+we will have several of them. For each of them, we will produce the best policy and the best
```

Comment:
"of them" moim zdaniem zbędne

Reply:
Ok, I've replaced "For each of them" with "For each run". Is this better?